README
Mail-OpenDKIM version 0.01
==========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by Nigel Horne

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Mail::OpenDKIM.

0.01  Thu Apr  7 12:09:43 2011
	- original version; created by h2xs 1.23 with options
		-A -n Mail::OpenDKIM



Makefile.PL
use 5.010000;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Mail::OpenDKIM',
    VERSION_FROM      => 'lib/Mail/OpenDKIM.pm', # finds $VERSION
    PREREQ_PM         => {
   	 'Error' => 0,
	 'Test::More' => 0.82,	# Needs new_ok
	 }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Mail/OpenDKIM.pm', # retrieve abstract from module
       AUTHOR         => 'Nigel Horne <nigel@kcilink.com>') : ()),
    LIBS              => ['-lopendkim'], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I. -I/usr/local/include', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/Mail/OpenDKIM/DKIM.pm
package Mail::OpenDKIM::DKIM;

use 5.010000;
use strict;
use warnings;

use Error;

use Mail::OpenDKIM;

our $VERSION = '0.01';

# Preloaded methods go here.

sub new {
	my ($class, $args) = @_;

	foreach(qw(dkimlib_handle)) {
		exists($$args{$_}) or throw Error::Simple("$class->new missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("$class->new undefined argument '$_'");
	}

	my $self = {
		_dkimlib_handle => $$args{dkimlib_handle},	# DKIM_LIB
		_dkim_handle => undef,	# DKIM
	};

	bless $self, $class;

	return $self;
}

sub dkim_sign
{
	my ($self, $args) = @_;

	if($self->{_dkim_handle}) {
		throw Error::Simple('dkim_sign called twice');
	}

	foreach(qw(id secretkey selector domain hdrcanon_alg bodycanon_alg sign_alg length)) {
		exists($$args{$_}) or throw Error::Simple("dkim_sign missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_sign undefined argument '$_'");
	}

	my $statp;

	$self->{_dkim_handle} = Mail::OpenDKIM::_dkim_sign($self->{_dkimlib_handle},
		$$args{id}, $$args{secretkey}, $$args{selector}, $$args{domain},
		$$args{hdrcanon_alg}, $$args{bodycanon_alg}, $$args{sign_alg},
		$$args{length}, $statp);

	return $statp;
}

sub dkim_header
{
	my ($self, $args) = @_;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_header called before dkim_sign');
	}
	foreach(qw(header len)) {
		exists($$args{$_}) or throw Error::Simple("dkim_header missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_header undefined argument '$_'");
	}

	return Mail::OpenDKIM::_dkim_header($self->{_dkim_handle}, $$args{header}, $$args{len});
}

sub dkim_eoh
{
	my $self = shift;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_eoh called before dkim_sign');
	}

	return Mail::OpenDKIM::_dkim_eoh($self->{_dkim_handle});
}

sub dkim_chunk
{
	my ($self, $args) = @_;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_chunk called before dkim_sign');
	}
	foreach(qw(chunkp len)) {
		exists($$args{$_}) or throw Error::Simple("dkim_chunk missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_chunk undefined argument '$_'");
	}

	return Mail::OpenDKIM::_dkim_chunk($self->{_dkim_handle}, $$args{chunkp}, $$args{len});
}

sub dkim_eom
{
	my $self = shift;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_eom called before dkim_sign');
	}

	return Mail::OpenDKIM::_dkim_eom($self->{_dkim_handle});
}

sub dkim_getsighdr_d
{
	my ($self, $args) = @_;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_getsighdr_d called before dkim_sign');
	}
	foreach(qw(initial buf len)) {
		exists($$args{$_}) or throw Error::Simple("dkim_getsighdr_d missing argument '$_'");
	}

	my $len;

	my $rc = Mail::OpenDKIM::_dkim_getsighdr_d($self->{_dkim_handle}, $$args{initial}, $$args{buf}, $len);

	if($rc == DKIM_STAT_OK) {
		$$args{len} = $len;
	}

	return $rc;
}

sub dkim_get_signer
{
	my $self = shift;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_get_signer called before dkim_sign');
	}

	return Mail::OpenDKIM::_dkim_get_signer($self->{_dkim_handle});
}

sub dkim_get_user_context
{
	my $self = shift;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_get_user_context called before dkim_sign');
	}

	my $rc = Mail::OpenDKIM::_dkim_get_user_context($self->{_dkim_handle});

	return ($rc == 0) ? undef : $rc;
}

sub dkim_set_final()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_final called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_final missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_final undefined argument '$_'");
	}

	return Mail::OpenDKIM::_dkim_set_final($self->{_dkimlib_handle}, $$args{func});
}

sub dkim_set_prescreen()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_prescreen called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_prescreen missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_prescreen undefined argument '$_'");
	}

	return Mail::OpenDKIM::_dkim_set_prescreen($self->{_dkimlib_handle}, $$args{func});
}

sub dkim_geterror
{
	my $self = shift;
	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_geterror called before dkim_sign');
	}

	return Mail::OpenDKIM::_dkim_geterror($self->{_dkim_handle});
}

sub dkim_free
{
	my $self = shift;

	unless($self->{_dkim_handle}) {
		throw Error::Simple('dkim_free called before dkim_sign');
	}

	my $rc = Mail::OpenDKIM::_dkim_free($self->{_dkim_handle});

	if($rc == DKIM_STAT_OK) {
		$self->{_dkim_handle} = undef;
	}

	return $rc;
}

sub DESTROY
{
	my $self = shift;

	if($self->{_dkim_handle}) {
		$self->dkim_free();
	}
}

1;

__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Mail::OpenDKIM::DKIM - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Mail::OpenDKIM::DKIM;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Mail::OpenDKIM, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head1 SUBROUTINES/Methods

=head2 new

For internal use by Mail::OpenDKIM only - do not call directly

=head2 dkim_sign

=head2 dkim_header

=head2 dkim_eoh

=head2 dkim_chunk

=head2 dkim_eom

=head2 dkim_getsighdr_d

=head2 dkim_get_signer

=head2 dkim_get_user_context

=head2 dkim_set_final

=head2 dkim_set_prescreen

=head2 dkim_geterror

=head2 dkim_free

=head2 EXPORT

All the function names and constants


=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Nigel Horne, E<lt>nigel@kcilink.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by MailerMailer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Mail/OpenDKIM.pm
package Mail::OpenDKIM;

use 5.010000;
use strict;
use warnings;

use Error;
require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Mail::OpenDKIM ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

use constant DKIM_CANON_SIMPLE => 0;	# RFC4871
use constant DKIM_CANON_RELAXED => 1;	# RFC4871
use constant DKIM_CANON_DEFAULT => DKIM_CANON_SIMPLE;

use constant DKIM_SIGN_RSASHA1 => 0;
use constant DKIM_SIGN_RSASHA256 => 1;

use constant DKIM_STAT_OK => 0;	# dkim.h
use constant DKIM_STAT_NORESOURCE => 6;
use constant DKIM_STAT_INVALID => 9;
use constant DKIM_STAT_NOTIMPLEMENT => 10;

use constant DKIM_FEATURE_DIFFHEADERS => 0;
use constant DKIM_FEATURE_DKIM_REPUTATION => 1;
use constant DKIM_FEATURE_PARSE_TIME => 2;
use constant DKIM_FEATURE_QUERY_CACHE => 3;
use constant DKIM_FEATURE_SHA256 => 4;
use constant DKIM_FEATURE_OVERSIGN => 5;
use constant DKIM_FEATURE_DNSSEC => 6;
use constant DKIM_FEATURE_RESIGN => 7;
use constant DKIM_FEATURE_ATPS => 8;

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	DKIM_CANON_RELAXED
	DKIM_CANON_SIMPLE
	DKIM_SIGN_RSASHA1
	DKIM_SIGN_RSASHA256

	DKIM_STAT_OK
	DKIM_STAT_NORESOURCE
	DKIM_STAT_INVALID
	DKIM_STAT_NOTIMPLEMENT

	DKIM_FEATURE_DIFFHEADERS
	DKIM_FEATURE_DKIM_REPUTATION
	DKIM_FEATURE_PARSE_TIME
	DKIM_FEATURE_QUERY_CACHE
	DKIM_FEATURE_SHA256
	DKIM_FEATURE_OVERSIGN
	DKIM_FEATURE_DNSSEC
	DKIM_FEATURE_RESIGN
	DKIM_FEATURE_ATPS
);

our $VERSION = '0.01';

require XSLoader;
XSLoader::load('Mail::OpenDKIM', $VERSION);

# Preloaded methods go here.
sub new {
	my $class = shift;

	my $self = {
		_dkimlib_handle => undef,	# DKIM_LIB
	};

	bless $self, $class;

	return $self;
}

sub dkim_init
{
	my $self = shift;

	if($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_init called more than once');
	}
	$self->{_dkimlib_handle} = _dkim_init();
	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_init failed to create a handle');
	}

	return $self;
}

sub dkim_close
{
	my $self = shift;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_close called before dkim_init');
	}
	_dkim_close($self->{_dkimlib_handle});
	$self->{_dkimlib_handle} = undef;
}

sub dkim_flush_cache
{
	my $self = shift;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_flush_cache called before dkim_init');
	}
	return _dkim_flush_cache($self->{_dkimlib_handle});
}

sub dkim_libfeature
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_libfeature called before dkim_init');
	}
	foreach(qw(feature)) {
		exists($$args{$_}) or throw Error::Simple("dkim_libfeature missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_libfeature undefined argument '$_'");
	}

	return _dkim_libfeature($self->{_dkimlib_handle}, $$args{feature});
}

sub dkim_sign
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_sign called before dkim_init');
	}
	foreach(qw(id secretkey selector domain hdrcanon_alg bodycanon_alg sign_alg length)) {
		exists($$args{$_}) or throw Error::Simple("dkim_sign missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_sign undefined argument '$_'");
	}
	require Mail::OpenDKIM::DKIM;

	my $dkim = Mail::OpenDKIM::DKIM->new({ dkimlib_handle => $self->{_dkimlib_handle} });

	my $statp = $dkim->dkim_sign($args);

	unless($statp == DKIM_STAT_OK) {
		throw Error::Simple("dkim_sign failed with status $statp");
	}

	return $dkim;
}

sub dkim_getcachestats
{
	my ($self, $args) = @_;

	return _dkim_getcachestats($$args{queries}, $$args{hits}, $$args{expired});
}

sub dkim_set_dns_callback()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_dns_callback called before dkim_init');
	}
	foreach(qw(func interval)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_dns_callback missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_dns_callback undefined argument '$_'");
	}

	return _dkim_set_dns_callback($self->{_dkimlib_handle}, $$args{func}, $$args{interval});
}

sub dkim_set_key_lookup()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_key_lookup called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_key_lookup missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_key_lookup undefined argument '$_'");
	}

	return _dkim_set_key_lookup($self->{_dkimlib_handle}, $$args{func});
}

sub dkim_set_policy_lookup()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_policy_lookup called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_policy_lookup missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_policy_lookup undefined argument '$_'");
	}

	return _dkim_set_policy_lookup($self->{_dkimlib_handle}, $$args{func});
}

sub dkim_set_signature_handle()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_signture_handle called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_signture_handle missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_signture_handle undefined argument '$_'");
	}

	return _dkim_set_signature_handle($self->{_dkimlib_handle}, $$args{func});
}

sub dkim_set_signature_handle_free()
{
	my ($self, $args) = @_;

	unless($self->{_dkimlib_handle}) {
		throw Error::Simple('dkim_set_signture_handle_free called before dkim_sign');
	}
	foreach(qw(func)) {
		exists($$args{$_}) or throw Error::Simple("dkim_set_signture_handle_free missing argument '$_'");
		defined($$args{$_}) or throw Error::Simple("dkim_set_signture_handle_free undefined argument '$_'");
	}

	return _dkim_set_signature_handle_free($self->{_dkimlib_handle}, $$args{func});
}

sub DESTROY
{
	my $self = shift;

	if($self->{_dkimlib_handle}) {
		$self->dkim_close();
	}
}

1;

__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Mail::OpenDKIM - Perl interface to the OpenDKIM library

=head1 SYNOPSIS

  use Mail::OpenDKIM;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Mail::OpenDKIM, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head1 SUBROUTINES/Methods

=head2 new

=head2 dkim_init

=head2 dkim_close

=head2 dkim_flush_cache

=head2 dkim_libfeature

=head2 dkim_sign

Returns an Mail::OpenDKIM::DKIM object

=head2 dkim_ssl_version

=head2 dkim_libversion

Static method.

=head2 dkim_getcachestats

Static method.

=head2 dkim_set_dns_callback

=head2 dkim_set_dns_callback

=head2 dkim_set_key_lookup

=head2 dkim_set_policy_lookup

=head2 dkim_set_signature_handle

=head2 dkim_set_signature_handle_free

=head2 EXPORT

All the function names and constants


=head1 SEE ALSO

http://www.opendkim.org/libopendkim/

=head1 AUTHOR

Nigel Horne, E<lt>nigel@kcilink.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by MailerMailer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut
